[WIP] Set Home Page

dtaylor113 opened the pull request:
Requires: https://github.com/openshift/origin-web-common/pull/260
Requires: https://github.com/openshift/origin-web-catalog/pull/575

Trello Card: https://trello.com/c/sg1xw7gG

**Adds 'Set Home Page' option to User dropdown:**
![image](https://user-images.githubusercontent.com/12733153/33559332-6f34e12e-d8da-11e7-913e-9f870bce1c2e.png)

**Set Home Page Modal: Multiple Projects**
![image](https://user-images.githubusercontent.com/12733153/33559422-b0ceff7a-d8da-11e7-9afd-e914e9d7a45e.png)

**Set Home Page Modal: One Project**
![image](https://user-images.githubusercontent.com/12733153/33622571-2e898a98-d9bc-11e7-90df-dfbad76efbe6.png)

**Set Home Page Modal: No Projects**
![image](https://user-images.githubusercontent.com/12733153/33559596-4f538b5c-d8db-11e7-9a84-fe663e393cfc.png)

**Notification When Home Page project No Longer Exists**
![image](https://user-images.githubusercontent.com/12733153/33622578-37071e74-d9bc-11e7-9b8e-2a0918b35403.png)
_* User is brought to projects list._

ncameronbritt commented:
@spadgett had an argument for defaulting to the landing page rather than the projects list page if the project had been deleted. Basically it was that for users on clusters with lots of projects, the shorter list of projects may be of more relevant to the user than the full list. I'm good with either. Interested in what Sam and @jwforres think.

For the radio buttons when there's only one project, it bothers me that the project select dropdown is disabled. I feel like that sends mixed messages--the radio button is active but the dropdown isn't. What if when there's only one project the third radio button just lists that project, with no dropdown? 
```
o Project Overview for My Project
```
Alternatively, I think having the dropdown active, even with one option, is preferable to the disabled dropdown.
cc @openshift/team-ux-review

dtaylor113 commented:
> .... it bothers me that the project select dropdown is disabled [when only one project].

Understood, however just FYI that is the standard behavior with the `<select-project>` component, which is also used in a few other wizards and modals.  When there is only one choice we default to it and disable the dropdown.  But I understand this may not be ideal when used in conjunction with radio buttons.

dtaylor113 commented:
@ncameronbritt when there is only one project, how should we display it without the dropdown:

    (  ) Overview Page for:
         My Second Project - my-second-project

....or something inline, ex:

    (  ) Overview Page for 'My Second Project' (my-second-project)

Looking for direction on how to display the project Display name and the project-name.  -thanks

ncameronbritt commented:
My preference is for inline. For the free tier of Online and probably other scenarios too, it will always be the case that as user only has access to a single project, so imho having it inline makes it appear more deliberate, as opposed to being an edge case we're designing around. 

Happy to hear other opinions though.

serenamarie125 commented:
Agreed on the inline suggestion

benjaminapetersen commented on app/scripts/controllers/util/error.js:
```diff
@@ -8,16 +8,18 @@
     switch(error) {
       case 'access_denied':
         $scope.errorMessage = "Access denied";
+        checkHomePagePreference(params.project_name);
```
This would typically throw an error w/our `jshint` conf, using a function before it is declared.  Do we still enforce that rule?

spadgett replied:
It would be nice if we can catch this somewhere before we redirect to the error page.

I think this will actually incorrectly show the message if you go to

```
/projects/my-project/pod/missing-pod
```

even if my-project exists

benjaminapetersen commented on app/scripts/controllers/util/error.js:
```diff
@@ -40,4 +42,29 @@ angular.module('openshiftConsole')
       $window.location.href = "/";
     };
+
+    function checkHomePagePreference(projectName) {
```
Curious if this function actually belongs within `HomePagePreference` rather than in `ErrorController`, even if it gets called here.  Should this module grow into a collection of error functions related to other modules?

benjaminapetersen commented on app/scripts/controllers/modals/setHomePageModal.js:
```diff
@@ -0,0 +1,74 @@
+    $scope.disableSaveButton = function() {
+      // if radiobuttons not changed
+      if ($scope.preferedHomePage === $scope.initialPreferedHomePage) {
+        if ($scope.preferedHomePage === 'project-overview') {
```
This `if` block is duplicated below on line 52, and a lot of nested `if` statements.  Care to dedupe w/function?  Can also prob eliminate the `// comments` with named fn calls.

benjaminapetersen commented on dist.java/java/404.html:
```diff
@@ -38,10 +38,10 @@ <h4>Connected to <span ng-bind="containerName"></span></h4>
       <div ng-include src="viewPartial"></div>
     </div>
 
```
Guessing you didn't change `dist.java`.

dtaylor113 replied:
Fixed

spadgett commented:
@dtaylor113 You have `dist.java` changes that shouldn't be there. I'd try

```
$ rm -rf dist.java
$ hack/install-assets
$ grunt build
```

spadgett commented on app/scripts/controllers/modals/setHomePageModal.js:
```diff
@@ -0,0 +1,74 @@
+'use strict';
+/* jshint unused: false */
```
Can we fix whatever is causing the warning instead of ignoring it?

spadgett commented on app/scripts/controllers/modals/setHomePageModal.js:
```diff
@@ -0,0 +1,74 @@
+ * @description
+ * # SetHomePageModalController
+ * Controller of the openshiftConsole
+ */
```
I don't think this comment is helpful. I understand a lot of our files have it, but I'd just take it out.

spadgett commented on app/scripts/controllers/modals/setHomePageModal.js:
```diff
@@ -0,0 +1,74 @@
+angular.module('openshiftConsole')
+  .controller('SetHomePageModalController', function ($scope, $uibModalInstance, HomePagePreference) {
+
+    $scope.initialPreferedHomePage = HomePagePreference.getHomePagePreference();
```
s/prefered/preferred

here and many places below

spadgett commented on app/scripts/controllers/modals/setHomePageModal.js:
```diff
@@ -0,0 +1,74 @@
+    $scope.selectedProject = null;
+    $scope.noProjects = false;
+
+    $scope.preSelectedProjectName = HomePagePreference.getProjectOverviewPage();
```
nit: "preselect" is one word, so `preselectedProjectName`

spadgett commented on app/scripts/controllers/modals/setHomePageModal.js:
```diff
@@ -0,0 +1,74 @@
+      // if radiobuttons not changed
+      if ($scope.preferedHomePage === $scope.initialPreferedHomePage) {
+        if ($scope.preferedHomePage === 'project-overview') {
+          if (!$scope.selectedProject) {
```
Can we just check `form.$pristine` instead to avoid all of this logic? The nested ifs are really hard to follow.

spadgett commented on app/scripts/controllers/util/error.js:
```diff
@@ -8,16 +8,18 @@
  */
 angular.module('openshiftConsole')
-  .controller('ErrorController', function ($scope, $window) {
+  .controller('ErrorController', function ($scope, $location, $window, $uibModal, HomePagePreference, NotificationsService) {
```
nit: `$uibModal` before `$window` to keep it alphabetized.

spadgett commented on app/scripts/controllers/util/error.js:
```diff
@@ -40,4 +42,29 @@ angular.module('openshiftConsole')
+            href: "",
+            label: "Set Home Page",
+            onClick: function() {
+              $uibModal.open({
```
Unfortunately this is going to leak memory because of https://github.com/openshift/origin-web-common/issues/97. I'm not sure what to suggest though until that issue is fixed.

spadgett replied:
> Unfortunately this is going to leak memory because of openshift/origin-web-common#97. I'm not sure what to suggest though until that issue is fixed.

Actually if you move the code that shows the notification into a service, I think it should be OK.

spadgett commented on app/scripts/directives/nav.js:
```diff
@@ -204,6 +204,19 @@ angular.module('openshiftConsole')
       restrict: 'EA',
       templateUrl: 'views/directives/header/header.html',
+      controller: function (AuthService, HomePagePreference) {
+        AuthService.withUser().then(function() {
```
I think this breaks the log out page? I'm not sure we can make this change since it will require every page to have an authenticated user.

Can we just check if `$rootScope.user` is set and not try to do anything if it's not.

spadgett commented on app/scripts/directives/nav.js:
```diff
@@ -204,6 +204,19 @@ angular.module('openshiftConsole')
+          HomePagePreference.init($rootScope.user);
+          switch(HomePagePreference.getHomePagePreference()) {
+            case 'project-list':
+              $location.path("/projects");
```
We can't unconditionally redirect. This will break direct linking to any URL in the web console! We should only use the value of path is empty.

I'm don't think the nav controller is the right place to make this check. Is it possible to do the check here?

https://github.com/openshift/origin-web-console/blob/master/app/scripts/app.js#L39-L59

Then add another `/catalog` route for the catalog that is used when you click the logo?

spadgett replied:
```js
Naviage.toProjectList();
```

dtaylor113 replied:
How would you recommend navigating to '/project/my-project/overview' in app.js's .config block when the preferred home page is set to a user's project overview?  -thanks

dtaylor113 replied:
This seems to work:
```
.run(function(HomePagePreference, Navigate) {
    var homePagePref = HomePagePreference.getHomePagePreference();
    if(homePagePref === "project-overview"){
      Navigate.toProjectOverview(HomePagePreference.getProjectOverviewPage());
    }
  })
```

spadgett replied:
> How would you recommend navigating to '/project/my-project/overview' in app.js's .config block when the preferred home page is set to a user's project overview?

Check the home page preference once in that block (if the user is logged in) and then 

```js
var homePagePath = <determine home page path...>
if (homePagePath) {
  $routeProvider.when('/', { redirectTo: homePagePath } );
} else {
  $routeProvider.when('/', {
    templateUrl: 'views/landing-page.html',
    controller: 'LandingPageController',
    reloadOnSearch: false
  });
}
```

dtaylor113 replied:
> I don't think the nav controller is the right place to make this check. Is it possible to do the check here?
> 
> https://github.com/openshift/origin-web-console/blob/master/app/scripts/app.js#L39-L59

Hi @spadgett,  the problem I'm running into is that the user is not logged in at the time app.js is executed.   I believed you and @jwforres indicated if getting the username was going to be an issue, then just store the homePagePref per browser and not use the userName in the key.   File a follow up issue to track.   Are we at that point? :-)

spadgett replied:
I'm OK with that.

One thing you could try, though, is to register an `AuthService.onUserChanged` callback in app.js and update the route to redirect if the user has a home page preference.

dtaylor113 replied:
Ok, so I have `AuthService.onUserChanged` working in app.js's .run block.   The only issue is that when user sets a new home page, clicking on the logo will still goto the old route/prev. home page until the user logs in again or reloads "/" in the browser (so that app.js .run/.config blocks are rerun to reset the `$routeParams`.

spadgett commented on app/scripts/directives/nav.js:
```diff
@@ -204,6 +204,19 @@ angular.module('openshiftConsole')
+              $location.path("/projects");
+              break;
+            case 'project-overview':
+              $location.path("/project/" + HomePagePreference.getProjectOverviewPage() + "/overview");
```
```js
Navigate.toProjectOverview(HomePagePreference.getProjectOverviewPage()));
```

spadgett commented on app/scripts/directives/util.js:
```diff
@@ -149,6 +149,22 @@ angular.module('openshiftConsole')
+  .directive('setHomePage', function($uibModal) {
+    return {
+      restrict: 'E',
+      replace: true,
```
`replace` is deprecated. We should avoid it.

dtaylor113 replied:
The only problem is that without the `replace` the HTML is:
```
<dropdown-menu>
  <li>
       <set-home-page>   <--- not 'replaced'
           <a href...>
```
Which breaks the CSS dropdown selectors, which many have:
`.dropdown-menu > li > a`

So 'Set Home Page' in the user dropdown menu isn't styled like the other menu options.

dtaylor113 replied:
Not sure we want to override all the `.dropdown-menu > li > a` selectors

spadgett replied:
The directive is simple enough I'd almost rather not have it be a directive. We could have a helper in a service to open the modal. Then all it is is a link with a one line click handler.

dtaylor113 replied:
I run into the same issue below where userDropdown.js is an extension where ng-click doesn't work, so I have to defer to using a directive.   The other `<a href=...`'s in the user and help dropdowns actually redirect to a URLs, but I don't think we want to do that for a modal.

benjaminapetersen replied:
Its a little too bad that the selector is `.dropdown-menu > li > a` instead of `.dropdown-menu  .link > a` or something that would provide a little more flexibility...

spadgett commented on app/scripts/directives/util.js:
```diff
@@ -149,6 +149,22 @@ angular.module('openshiftConsole')
+      replace: true,
+      template: '<a href="">Set Home Page</a>',
+      link: function($scope, element) {
+        element.bind('click', function() {
```
Why not use `ng-click`?

dtaylor113 replied:
The directive is being used in an extension.  I did try ng-click in the nav-menu controller but the extensions seem to not trigger the ng-click function in the controller.

spadgett replied:
OK I see the problem and it's what we do for copy login link. I'm OK with it.

benjaminapetersen replied:
Is a bug with extensions?  I believe there is a way to `ng-click`, but it's been a good while since I've looked @ the code.  Doesn't need to hold up this PR, but I can def take a look @ it if its causing trouble.

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,37 @@
+angular.module("openshiftConsole")
+  .service("HomePagePreference", function(){
+
+    var HOME_PAGE_PREF_KEY = "origin/home-page-pref/";
```
We use `openshift/` as the prefix elsewhere. Let's be consistent here.

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,37 @@
+    var HOME_PAGE_PREF_KEY = "origin/home-page-pref/";
+
+    var init = function(user) {
+      HOME_PAGE_PREF_KEY += user.metadata.name;
```
This could append the user name multiple times if called more than once. I think it's better to add a method to AuthService that gets the current user (without triggering a new login).

```js
var HOME_PAGE_KEY_PREFIX = "openshift/home-page-preference/";
var getHomePagePreference = function() {
  var user = AuthService.getUser(); // add to method
  if (!user) {
    return null;
  }
  var key = HOME_PAGE_KEY_PREFIX + user;
  // ...
};

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,37 @@
+    };
+
+    var getHomePagePreference = function() {
+      var homePagePref = JSON.parse(localStorage.getItem(HOME_PAGE_PREF_KEY));
```
we should have a try/catch to for JSON parse errors just in case.

benjaminapetersen replied:
or use lodash `_.attempt()`

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,37 @@
+    };
+
+    var getProjectOverviewPage = function() {
+      var homePagePref = JSON.parse(localStorage.getItem(HOME_PAGE_PREF_KEY));
```
try/catch

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,37 @@
+
+    var getHomePagePreference = function() {
+      var homePagePref = JSON.parse(localStorage.getItem(HOME_PAGE_PREF_KEY));
+      return homePagePref ? homePagePref.type : "catalog-home";
```
```js
return _.get(homePagePref, 'type', 'catalog-home');
```

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,48 @@
+  </div>
+  <div class="modal-body indent">
+    <p>
+      <div class="title">Set your preferred home page.  This is the page you will see at login.</div>
```
It's not necessarily the page you see when you login. It's the page if there's no URL path. We need to let users link directly to other pages.

dtaylor113 replied:
@ncameronbritt, could you please re-wordsmith this?  -thanks

ncameronbritt replied:
I feel like just removing that second sentence covers it. I believe that users would understand what a home page means.

spadgett commented on app/scripts/controllers/util/error.js:
```diff
@@ -40,4 +42,29 @@ angular.module('openshiftConsole')
+          id: "invalid-home-page-preference",
+          type: "warning",
+          message: "Home page set to invalid project.",
+          details: "Project '" + projectName + "' no longer exists.",
```
"no longer exists or you do not have access to it."

spadgett commented on app/scripts/controllers/util/error.js:
```diff
@@ -40,4 +42,29 @@ angular.module('openshiftConsole')
+        NotificationsService.addNotification({
+          id: "invalid-home-page-preference",
+          type: "warning",
+          message: "Home page set to invalid project.",
```
"invalid" doesn't really seem right. Maybe

> Home page project not found.
> Project my-project no longer exists or you do not have access to it.

spadgett commented:
Should clicking on the OpenShift logo take you to the home page or the catalog page? If it takes you to your home page, do we need another way to get to the catalog page?

dtaylor113 commented:
> Should clicking on the OpenShift logo take you to the home page or the catalog page? If it takes you to your home page, do we need another way to get to the catalog page?

With the latest code.  If homePage is set to project-list, then clicking on the logo brings you to the project list.   If homePage is undefined, 'catalog-home', or 'project-overview' -clicking on the logo brings you to the browse catalog landing page.  ...and that's only because I'm still trying to figure out how to default "/" to a specific project overview in app.js.

ncameronbritt commented:
That's a good question @spadgett. On the one hand, users have the expectation that clicking the logo will take them "home". But on the other hand, with Jeff's browse in project stuff, we potentially lose any ability to get to that landing page if OpenShift logo doesn't go there. And the application IA can look different depending on whether or not a user has specified something other than the landing page to be their home page.

It didn't occur to me that users already had the ability to specify where they land upon logging in by using bookmarks. Right now it would be easy enough to bypass the landing page simply by bookmarking another page. Is the idea with this design that *whenever* users log in they end up on their home page? If I have bookmarked different pages that are important to me, will those bookmarks still work if I have to log in?

dtaylor113 commented:
Hi @spadgett & @ncameronbritt ,  I believe I have addressed all your review issues.  
I updated the screenshots in this PR's Description.

Only two issues remain:

1. After user sets a new home page, clicking on the logo will redirect to prev. home page setting until user reloads "\" or logs out/in.   We could have the Set Home Page modal on Save reload "/", not sure if that's good user experience though as they may initiate the Set Home Page modal from any location and may not want to goto the home page pref upon Saving.

2. @spadgett,  still using the `<set-home-page>` directive with `replace: true`.  Can you elaborate on how you might launch the modal from userDroppdown.js (extension) without using a directive?  Can't use `ng-click` as userDroppdown is an extension.  -thanks

spadgett commented on dist/styles/fonts/openshift-logos-icon.svg:
```diff
@@ -9,8 +9,8 @@
-	"minorVersion": 3,
-	"version": "Version 1.3",
+	"minorVersion": 2,
+	"version": "Version 1.2",
```
Make sure you run `bower install` since this dependency was updated.

beanh66 commented:
@ahardin-rh FYI, just wanted you to be aware of this feature!

beanh66 commented:
@dtaylor113 Per the comment from @spadgett, I agree it is strange if we don't allow users any way to get back to the catalog landing page. In the original design we had said the logo in the masthead should still take users to the catalog landing page and "set home page" would only alter the page that appears on initial login.

dtaylor113 commented:
@beanh66 , ok, I'll update it so that the logo always brings user to the Catalog.  -thanks

serenamarie125 commented:
@dtaylor113 I think that makes sense for now, otherwise the user never has the opportunity to access the catalog with SaaS offerings.  Thanks!

dtaylor113 commented:
Updated so that update it so that clicking on logo always brings user to the Catalog (unless catalog is disabled via config.js)

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+      <div class="title">Set your preferred home page.</div>
+      Note: This setting is browser specific and will not be maintained across browsers.
+    </p>
+    <ng-form name="form">
```
```html
<form name="form">
```

Give it a more specific name, though. Since this is a dialog shown on any page, the generic name `form` might already be used.

dtaylor113 replied:
Fixed

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+  </div>
+  <div class="modal-footer">
+      <button class="btn btn-lg btn-primary" type="button"
+              ng-disabled="(preferredHomePage === 'project-overview' && form.$invalid) || form.$pristine"
```
It would be better to update `select-project` to take an optional required attribute that is set to true only when `preferredHomePage === 'project-overview'`. Then this changes to

```
ng-disabled="form.$invalid || form.$pristine"
```

dtaylor113 replied:
Hi @spadgett, sorry, I little confused on "...update select-project to take an optional required attribute..",  Can you please elaborate on 'optional required attribute`? -Thanks

spadgett replied:
Sorry I phrased that poorly. I mean something like

```html
<select-project is-required="preferredHomePage === 'project-overview'" ...>
```

Then `form.$invalid` will be the correct value regardless of `preferredHomePage`

spadgett commented on app/views/overview/_service-bindings.html:
```diff
@@ -19,7 +19,7 @@
     <span>You must have a bindable service in your namespace in order to create bindings.</span>
     <div>
-      <a href="./">Browse Catalog</a>
+      <a href="./browse-catalog">Browse Catalog</a>
```
```html
<a ng-href="{{project | catalogURL}}">
```

dtaylor113 replied:
Hmm...I don't see `project` defined in `/app/scripts/directives/overview/serviceBindings.js`

dtaylor113 commented on app/scripts/app.js:
```diff
@@ -35,6 +35,14 @@ angular
     'webCatalog'
   ])
+  .run(function(AuthService, HomePagePreference) {
+    HomePagePreference.gotoHomePage();
```
yes it does :-) sorry.  So we only want to go to the home page when path is empty.  `$location.$$url` seems to contain this info; path minus the baseHref:

```
When https://localhost:9000/dev-console/project/myproject/browse/stateful-sets/
$location.$$url =  /project/myproject/browse/stateful-sets/

When https://localhost:9000/dev-console/
$location.$$url = "/"
```
Seem reasonable? Unless there is a helper function which already tests this.
Thanks,
Dave

dtaylor113 replied:
Added code to only go to home page when path is empty.

dtaylor113 replied:
Arrggg...now if you set your home page pref to 'project-list' and log out and back in, it doesn't redirect to '/projects' even though the $location.path('projects') is called.   

As I test I have the code below, but even though `$location.path('projects')` is called during `onUserChanged` browser doesn't redirect to the Project List.

```
.run(function(AuthService, HomePagePreference, $location) {
    if ($location.$$url === "/") {
      HomePagePreference.gotoHomePage();
    }
    AuthService.onUserChanged(function(user) {
      if (user) {
        //HomePagePreference.gotoHomePage();
        $location.path('projects');
      }
    });
  })
```

It's weird because when log in to project which no longer exists, you do get the toast notification, **and a redirect to Project List!**.   Log in/out works fine with Catalog(Home) and Project Overview set as home page pref, just not the project-list.

dtaylor113 replied:
Ok, had to add a `$timeout` [here](https://github.com/openshift/origin-web-console/pull/2562/files#diff-f9201982127fd8efc3606dcbb9fa087fR110) in order to navigate.

benjaminapetersen commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,130 @@
+      try {
+        homePagePref = JSON.parse(localStorage.getItem(key));
+      } catch (e) {
+        return 'catalog-home';
```
Yeah, swallowing the error entirely will probably be really frustrating to debug at some point.

dtaylor113 replied:
Fixed

dtaylor113 commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,130 @@
+      var deferred = $q.defer();
+      var projects, projectNames;
+
+      ProjectsService.list().then(function(projectData) {
```
duh..yes..sorry :-)

dtaylor113 replied:
Fixed

dtaylor113 commented on app/scripts/app.js:
```diff
@@ -57,6 +90,11 @@ angular
 
     $routeProvider
       .when('/', landingPageRoute)
+      .when('/browse-catalog', {
```
I actually took this out of my PR and just limited this PR to only going to home page preference upon login.  I reverted all my previous changes to "Browse Catalog" links in favor of #2563.

dtaylor113 commented on app/views/directives/resource-service-bindings.html:
```diff
@@ -19,7 +19,7 @@ <h3>Bindings</h3>
     <span>You must have a bindable service in your namespace in order to create bindings.</span>
     <div>
-      <a href="./">Browse Catalog</a>
+      <a href="./browse-catalog">Browse Catalog</a>
```
Fixed

dtaylor113 commented on app/views/overview.html:
```diff
@@ -12,7 +12,7 @@ <h2>Get started with your project.</h2>
           </p>
           <p class="gutter-top">
-            <a ng-if="overview.catalogLandingPageEnabled" href="./" class="btn btn-lg btn-primary">
+            <a ng-if="overview.catalogLandingPageEnabled" href="./browse-catalog" class="btn btn-lg btn-primary">
```
Reverted this change in favor of what was done in #2563

dtaylor113 commented on app/scripts/app.js:
```diff
@@ -35,8 +58,18 @@ angular
+  .run(function(AuthService, HomePagePreference, Navigate) {
+    setHomePage(HomePagePreference, Navigate);
+    AuthService.onUserChanged(function(user) {
+      if (user !== null) {
```
Fixed

dtaylor113 commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,130 @@
+      try {
+        homePagePref = JSON.parse(localStorage.getItem(key));
+      } catch (e) {
+        return null;
```
Fixed

dtaylor113 commented on app/views/directives/header/header.html:
```diff
@@ -40,7 +40,7 @@
       <li ng-if-end role="menuitem"><a ng-href="project/{{currentProjectName}}/create?tab=fromFile">Import YAML / JSON</a></li>
 
-      <li ng-if-start="catalogLandingPageEnabled" role="menuitem"><a href="./">Browse Catalog</a></li>
+      <li ng-if-start="catalogLandingPageEnabled" role="menuitem"><a href="./browse-catalog">Browse Catalog</a></li>
```
Reverted, not using `/browse-catalog` or `catalog` route

spadgett commented:
@dtaylor113 I feel like we might be making this harder on ourselves than we need to. Is it possible to skip checking the user and just set up a redirect using...?

```js
$routeProvider.when('/', {
  redirectTo: myHomePage
});
```

dtaylor113 commented:
> $routeProvider.when('/', {
>   redirectTo: myHomePage
> });

What is `myHomePage`?  And in order to determine `myHomePage` won't we need to check the user's home page preference?

spadgett commented:
> What is myHomePage? And in order to determine myHomePage won't we need to check the user's home page preference?

Yes, but if we don't tie it to a specific user ID, it's just a check to localStorage

dtaylor113 commented:
> Yes, but if we don't tie it to a specific user ID, it's just a check to localStorage
>> Ok, so your suggesting doing away with the user specific setting and just having a home page per browser (regardless of what user is logged in).

If so, then the only edge case maybe if userOne sets home to a project, then userTwo logs in but doesn't have access to said project, so userTwo will get the 'project no longer exists' toast notification where they can set home page to a project they have access to...so two users on same browser might keep stepping on each others toes.

dtaylor113 commented:
Doesn't a user have to be logged in in order to do the `verifyProjectOverviewPage()` check where we do a `ProjectsService.get()..which uses....AuthService.withUser().then('...`  ?

Also having trouble using a promise in the `redirectTo`:

```
$routeProvider
      .when('/', {
        redirectTo: function() {
          HomePagePreferenceProvider.$get().getHomePageURL().then(function (url) {
            return url;
          });
          //returns before the .then(...)
        }
      })
```

spadgett commented:
@dtaylor113 We can try `resolveRedirectTo` instead of `redirectTo`, but it means we'd have to update to AngularJS 1.6.

https://code.angularjs.org/1.6.1/docs/api/ngRoute/provider/$routeProvider

dtaylor113 commented:
> @dtaylor113 We can try resolveRedirectTo instead of redirectTo, but it means we'd have to update to AngularJS 1.6.

I can try that if you don't like the `.run` block.

spadgett commented on app/scripts/app.js:
```diff
@@ -35,6 +35,16 @@ angular
     'webCatalog'
   ])
+  .run(function(AuthService, HomePagePreference, $location) {
+    if ($location.$$url === "/") {
```
`$$url` is an internal angular property that we should avoid using.

spadgett commented on app/scripts/app.js:
```diff
@@ -35,6 +35,16 @@ angular
+    }
+    AuthService.onUserChanged(function(user) {
+      if (user) {
+        HomePagePreference.gotoHomePage();
```
It still seems like this could break direct linking? I think we need to avoid the `run()` block.

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,131 @@
+        });
+      } else if (homePagePref === "project-list") {
+        // navigate on next digest cycle
+        $timeout(function(){
```
Why is the timeout needed? Because we're calling from a `run()` block? I'd rather not have this in the service, but have it in the run block. But I'm guessing we're be able to remove this if we switch to `redirectTo`

spadgett commented on app/styles/_core.less:
```diff
@@ -978,6 +978,16 @@ a.disabled-link {
     margin-top: 0px;
     margin-bottom: 10px;
   }
+  .indent {
```
These styles are specific to the home page dialog and shouldn't be in core.less or have such generic names.

spadgett commented on app/styles/_core.less:
```diff
@@ -978,6 +978,16 @@ a.disabled-link {
+  }
+  .select-project-container {
+    padding-left: 20px;
+    width: 70%;
```
How does this look at mobile?

Better to give it a max width in px instead of a percentage width. Then it can be 100% at mobile.

spadgett commented on dist/styles/fonts/openshift-logos-icon.svg:
```diff
@@ -35,7 +35,6 @@
 <glyph unicode="&#xe907;" glyph-name="rh-integration" data-tags="logo-rh-integration" d="M314.163 645.325c-2.102-0.462-4.516-0.727-6.992-0.727-8.313 0-15.928 2.983-21.835 7.938l0.053-0.043c-0.309 0.078-0.664 0.122-1.030 0.122-2.432 0-4.403-1.971-4.403-4.403 0-0.079 0.002-0.157 0.006-0.235v0.011c0.438-10.625 9.157-19.072 19.85-19.072 7.325 0 13.724 3.964 17.168 9.864l0.051 0.094c0.355 0.621 0.564  ...
 <glyph unicode="&#xe908;" glyph-name="rh-tomcat" data-tags="logo-rh-tomcat" d="M559.104 343.552l-47.104-47.104-47.104 47.104c-4.096 3.072-9.216 3.072-13.312 0-4.096-4.096-5.12-10.24-1.024-14.336l51.2-51.2v-48.128c0-6.144 4.096-10.24 10.24-10.24s10.24 4.096 10.24 10.24v48.128l51.2 51.2c4.096 4.096 4.096 10.24 0 14.336-4.096 3.072-10.24 3.072-14.336 0zM382.976 425.472l-9.216 9.216c-4.096 3.072-9.21 ...
 <glyph unicode="&#xe909;" glyph-name="sso" data-tags="logo-sso" d="M638.566 674.304h0.614v0.614zM741.069 574.054h-0.614v-0.614zM743.322 585.318l-17.613 65.638c-1.349 4.382-4.731 7.765-9.017 9.088l-0.096 0.026-65.638 17.613c-1.095 0.379-2.356 0.597-3.669 0.597-6.391 0-11.571-5.181-11.571-11.571 0-1.313 0.219-2.574 0.621-3.751l-0.024 0.081 45.363-45.363 47.002-46.797c1.095-0.379 2.356-0.597 3.669-0 ...
-<glyph unicode="&#xe90a;" glyph-name="openshift" data-tags="logo-openshift" d="M300.992 496.168l-162.604-59.173c2.076-26.073 6.575-51.829 12.881-77.054l154.469 56.242c-5.021 26.108-6.801 53.029-4.746 79.985zM1019.614 675.975c-11.335 23.39-24.456 46.002-39.654 67.302l-162.576-59.166c18.919-19.357 34.795-41.101 47.775-64.364l154.455 56.228zM661.892 722.027c33.82-15.784 63.121-37.316 87.831-62.563l1 ...
```
Your PR has changes to logos that are unrelated.

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,131 @@
+    };
+
+    var notifyInvalidProjectHomePage = function(projectName) {
+      clear();
```
This will `clear` your preference on any error. So if you lose network connectivity or something and the project is fine, it still clears your preference, which it shouldn't.

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,131 @@
+      ProjectsService.get(projectOverviewPage).then(function() {
+        deferred.resolve(true);
+      }, function() {
+        deferred.resolve(false);
```
We really need to check the status code. If it's not 403 or 404, the project might still exist.

dtaylor113 commented:
Closing this WIP PR, too many comments.  I'll create another PR which hopefully is closer to the final solution.
